Web Inspector: Page/frame targets in the same web process should share InjectedScriptManager

the-chenergy opened the pull request:
#### 165749535eb7299332dfd0d2b5d718d7465c8262
<pre>
Web Inspector: Page/frame targets in the same web process should share InjectedScriptManager
<a href="https://rdar.apple.com/164199657">rdar://164199657</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302120">https://bugs.webkit.org/show_bug.cgi?id=302120</a>

Reviewed by BJ Burg.

Having many InjectedScriptManagers in the same web process and same VM
will cause problems when we start implementing agents in the frame
target and move them from the page. Make the WebInjectedScriptManager
ref-counted so that the page target (InspectorController) can share it
with its frame target (FrameInspectorController).

No new tests. No expected changes in noticeable behavior.

* Source/WebCore/inspector/FrameInspectorController.cpp:
(WebCore::FrameInspectorController::FrameInspectorController):
(WebCore::FrameInspectorController::createLazyAgents):
(WebCore::FrameInspectorController::disconnectFrontend):
(WebCore::FrameInspectorController::inspectedFrameDestroyed):
* Source/WebCore/inspector/FrameInspectorController.h:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::InspectorController):
* Source/WebCore/inspector/InspectorController.h:
* Source/WebCore/inspector/WebInjectedScriptManager.cpp:
(WebCore::WebInjectedScriptManager::create):
* Source/WebCore/inspector/WebInjectedScriptManager.h:
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::WorkerInspectorController):
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/inspector/agents/WebConsoleAgent.cpp:
(WebCore::WebConsoleAgent::frameWindowDiscarded):

Canonical link: <a href="https://commits.webkit.org/304471@main">https://commits.webkit.org/304471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a4b904bab7d75705de642ac5bbd4ab96f854b9b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134671 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7126 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142195 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86610 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/015ff881-201b-4bfd-be4b-be289228dc8b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7728 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6979 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102928 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c96e05fd-6b65-41a3-ad47-88239a302328) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137618 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120703 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83733 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/00001783-e846-4e5f-b08c-5869a97b1f9f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5278 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2891 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2785 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38844 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144887 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6800 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39425 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111323 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6874 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5701 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111618 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5113 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116977 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60684 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20910 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6851 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35168 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6656 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70432 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6892 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6765 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso reviewed (changes requested):
as we discussed in person at the WebKit Contributors Meeting last week, i don't think this overall approach is the right one

as a general concept, there should never be more than one `InspectorEnvironment` for each JS thread since we don't want multiple `Debugger` to exist and be able to pause (as that would require syncing state between them)

the current design of the `Target` system (which in hindsight probably could've been named better) is really just about ensuring that the protocol JSON messages are passed to the right place before being unpacked and handled by some agent that's there

for example, today Web Inspector will connect to a `WebPageProxy` in the UIProcess as a way of "multiplexing" messages to the right `WebPage` in the right WebProcess if a "process swap on navigation" occurs

with site-isolation, i think all we really want to do is have another target in the UIProcess that's able to send messages to the WebProcess for the cross-origin `<iframe>`

put another way, having a `FrameInspectorController` and a (`Page`)`InspectorController` at the same time is more complex and not any better than the flow we have right now

today, regardless of whether a webpage has same-origin or cross-origin `<iframe>` it will all be handled by the same `InspectorDOMAgent`

i think the only thing that should change with site-isolation is when it's cross-origin, and in that case it's really just that we have to route the protocol JSON message to a different WebProcess where it'll be handled by a different `InspectorDOMAgent`

there may be some logic needed to make sure it's created at the right time and owned by the right object, but beyond that i think everything else should work pretty much the same as before (though the frontend will likely need some extra love to fix these new "gaps" where we go from one target to a different target)

i apologize for not sharing this in earlier reviews, but i didn't really have all the details then. happy to chat more here or offline if that'd help :)

burg reviewed (commented):
Re: this patch, InjectedScriptManager is part of InspectorEnvironment, which is assumed to outlive any particular agent. It is also true that Page will outlast the Frames. I think what we actually need here is UniqueRef for owning m_injectedScriptManager + CheckedRef when we hand it off to agents.

The bigger picture for frame targets is not just about implementability, but about what the API should be going forward for interacting with pages, frames, workers, etc. I've been writing up an explainer with my conclusions and will share that out before Thanksgiving break.

However, for the purposes of adopting saferCPP, we should continue iterating on this patch.

the-chenergy committed:
Web Inspector: Page/frame targets in the same web process should share InjectedScriptManager
rdar://164199657
https://bugs.webkit.org/show_bug.cgi?id=302120

Reviewed by BJ Burg.

Having many InjectedScriptManagers in the same web process and same VM
will cause problems when we start implementing agents in the frame
target and move them from the page. Make the WebInjectedScriptManager
ref-counted so that the page target (InspectorController) can share it
with its frame target (FrameInspectorController).

No new tests. No expected changes in noticeable behavior.

* Source/WebCore/inspector/FrameInspectorController.cpp:
(WebCore::FrameInspectorController::FrameInspectorController):
(WebCore::FrameInspectorController::createLazyAgents):
(WebCore::FrameInspectorController::disconnectFrontend):
(WebCore::FrameInspectorController::inspectedFrameDestroyed):
* Source/WebCore/inspector/FrameInspectorController.h:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::InspectorController):
* Source/WebCore/inspector/InspectorController.h:
* Source/WebCore/inspector/WebInjectedScriptManager.cpp:
(WebCore::WebInjectedScriptManager::create):
* Source/WebCore/inspector/WebInjectedScriptManager.h:
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::WorkerInspectorController):
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/inspector/agents/WebConsoleAgent.cpp:
(WebCore::WebConsoleAgent::frameWindowDiscarded):

Canonical link: https://commits.webkit.org/304471@main
